keras/backend/tensorflow_backend.py

```diff
@@ -2486,13 +2486,33 @@ def __init__(self, inputs, outputs,
             raise ValueError('Some keys in session_kwargs are not '
                              'supported at this '
                              'time: %s', session_kwargs.keys())
+        self._extra_fetches_info = dict()
         self._callable_fn = None
-        self._merged_summaries = None
         self._feed_arrays = None
         self._feed_symbols = None
         self._symbol_vals = None
         self._session = None
 
+    def _add_fetch(self, fetch, fetch_callback):
+        fetch_index = len(self.fetches)
```

Comment:
I'm against using `len(self.fetches)` as key because `fetches` is a property that can be altered at runtime by other callbacks, breaking the uniqueness of `fetch_index`. For example:
```
2. You add `Tensorboard#merged` to it
3. Some "cleaning" callback `A` sets `test_function.fetches = []`
3. Some other callback `B` calls `test_function._add_fetch(...)`, overriding Tensorboard's callback
4. Tensorboard calls _remove_fetch, removing `B`'s callback.
```
Why not just `self._extra_fetches_info[fetch] = fetch_callback`? This seems more reasonable because `fetch` is an operation owned by the caller.

Reply:
Using the fetch as the key, the (current) index of the fetch in `fetches` could be found to retrieve the right result.
In that case, adjusting `fetches` without refreshing the callable would be a problem.

Reply:
Ideally, callbacks would use these new functions in the future to manage fetches without interfering with other callbacks in unforseen ways.

The current code could break:
- \>= TF1.8
  - using a callback using this new functionality before old code writing to `fetches` in a way that alters the existing indices, followed by a refresh of the callable (for example another callback using this new functionality)
- < TF1.8
  - in combination with old code writing to `fetches` in a way that alters the existing indices after a callback using this new functionality

Using the extra fetches as dict keys (and adding these extra fetches to `fetches`), this would become:
- \>= TF1.8
  - using a callback using this new functionality before old code writing to `fetches` in a way that removes the extra fetch
- < TF1.8
  - in combination with old code writing to `fetches` in a way that removes the extra fetch after a callback using this new functionality

With both implementations, adding the callbacks using the new functionality last in the callback array would solve the problem.

To prevent interference with `fetches`, using the extra fetches as the dict keys (but not adding these extra fetches to `fetches`), the dict keys could be added [here](https://github.com/hermansje/keras/blob/Tensorboard_callback_refactor/keras/backend/tensorflow_backend.py#L2549) and [here](https://github.com/hermansje/keras/blob/Tensorboard_callback_refactor/keras/backend/tensorflow_backend.py#L2620).
That would also 'hide' the extra fetches the function does. `fetches` are then those passed by the user, not those done by the function.

Reply:
> Ideally, callbacks would use these new functions in the future to manage fetches without interfering with other callbacks in unexpected ways.

I understand the necessity, but I disagree with this approach. Given there's a public property `Function#fetches` and a private method `_add_fetch`, i fear users will be automatically driven to use the former, causing a lot of confusion in future extensions.

> In that case, adjusting fetches directly (without refreshing the callable) would be a problem.

True, but there's a way to keep track of changes in `fetches` and update the function -- the same way it's done with `_feed_symbols` --:

We keep 2 fetches:

```py
    def __init__(...):
        ....
        self.fetches = [tf.identity(x) for x in self.fetches]
        ...
        self._callable_fn = None
        self._feed_arrays = None
        self._feed_symbols = None
        self._fetches = None
```
Update the callback every time `fetches` is updated:
```py
def _call(...):
        ....
        # Refresh callable if anything has changed.
        if (self._callable_fn is None or
            feed_arrays != self._feed_arrays or
            symbol_vals != self._symbol_vals or
            feed_symbols != self._feed_symbols or
            self.fetches != self._fetches or         # <--- Here!
            session != self._session):
```
Update the **true** fetches:
```py
    def _make_callable(...):
        ...
        self._callable_fn = callable_fn
        self._feed_arrays = feed_arrays
        self._feed_symbols = feed_symbols
        self._symbol_vals = symbol_vals
        self._fetches = list(self.fetches)         # <--- Here!
```
Callback `_fetches` instead of `fetches`, as they haven't been updated since the function was last called:
```
    def _call(...):
        ...
        for f, output in zip(self._fetches, fetched):
            if f in self._extra_fetches_info:
                 self._extra_fetches_info[f](output)
```
`Tensorboard` would be written like this:
```py
        if not self.histogram_freq and epoch % self.histogram_freq:
            fetch_callback = lambda res: self.writer.add_summary(res, epoch)
            self.model.test_function.fetches += [self.merged]
            self.model.test_function._extra_fetches_info[self.merged] = fetch_callback
```

All written code up to this PR would be kept as it is. All new callbacks can just `fn.fetches.pop/append`.
Also, `_extra_fetches_info` should probably become public, and maybe something containing callbacks in the name?

Problem solved.

Reply:
I updated the code to use `_fetches` and `fetch_callbacks`.